Add Spark SQL support

gengliangwang opened the pull request:
# Add Spark SQL support 
* Add Spark SQL support. It can connect to Spark via building a local/remote SparkSession.
* Include a notebook example

I tried some complicated queries (window function, table joins), and the tool works well. 
Compared to the [Spark Dataframe agent](https://python.langchain.com/en/latest/modules/agents/toolkits/examples/spark.html), this tool is able to generate queries across multiple tables.

gengliangwang committed:
Spark SQL Agent

gengliangwang commented:
Note: There was an approach based on `SQLDatabase`. But @dev2049 suggests not inheriting from SQLDatabase.
https://github.com/hwchase17/langchain/pull/4381

gengliangwang committed:
fix lint

gengliangwang committed:
revise wording

skcoirz committed:
Merge remote-tracking branch 'upstream/master' into spark_sql_no_subsclass

skcoirz committed:
fix lint error

skcoirz committed:
fix public api assertion

skcoirz committed:
specify the dialect. to make it more certain for LLM

skcoirz commented on langchain/tools/spark_sql/tool.py:
```diff
@@ -0,0 +1,151 @@
+                    template=QUERY_CHECKER, input_variables=["query"]
+                ),
+            )
+
```
if we force users to use the default prompt, I think it makes sense here and then I would change the error message below to be more specific.

skcoirz committed:
updated error message a little bit to match the requirement

gengliangwang commented:
@skcoirz Thanks for help updating this one!

skcoirz commented:
> @skcoirz Thanks for help updating this one!

yeah, sure thing! I tested this. The new query checker is really powerful! It solved the previous concern of AnalysisException. Thank you so much for adding this! During the test, I noticed a few more opportunities. I have added them to our spreadsheet. Happy to chat more when you have time! Have a good weekend! :D

skcoirz committed:
added another checker as LLM output sometimes fails to follow the format requirement

skcoirz committed:
format

skcoirz committed:
Merge remote-tracking branch 'upstream/master' into spark_sql_no_subsclass

skcoirz committed:
add memory support to spark sql agent

skcoirz committed:
add import for pyexception

skcoirz committed:
finished testing. all good

gengliangwang commented on langchain/agents/agent_toolkits/spark_sql/base.py:
```diff
@@ -0,0 +1,110 @@
+from langchain.spark_sql import SparkSQL
+
+
+def create_spark_analytics_agent_verified(
```
@skcoirz Having two methods for creating agents looks confusing.

gengliangwang commented on langchain/agents/agent_toolkits/spark_sql/base.py:
```diff
@@ -0,0 +1,110 @@
+    which is verified during development.
+    """
+    spark_sql = SparkSQL(schema=schema)
+    llm = ChatOpenAI(temperature=0, model_name="gpt-4")
```
Let's not bind gpt-4 here. Langchain is supposed to be general.

gengliangwang replied:
@skcoirz I am reverting the related changes. You can still find the code in the git history.

skcoirz replied:
sure, I’m good with it. There is a tradeoff. We can chat more later.

gengliangwang committed:
revise comment

skcoirz commented:
Moved the rest new features to a new PR on top of branch. (https://github.com/hwchase17/langchain/pull/4672)

gengliangwang commented:
cc @vowelparrot @hwchase17 could you review this one? The new agent is helpful for the Apache Spark community.

hwchase17 committed:
bump version to 173 (#4910)

eyurtsev committed:
Load specific file types from Google Drive (issue #4878) (#4926)

# Load specific file types from Google Drive (issue #4878)
Add the possibility to define what file types you want to load from
Google Drive.
 
```
 loader = GoogleDriveLoader(
    folder_id="1yucgL9WGgWZdM1TOuKkeghlPizuzMYb5",
    file_types=["document", "pdf"]
    recursive=False
)
```

Fixes ##4878

## Who can review?
Community members can review the PR once tests pass. Tag
maintainers/contributors who might be interested:
DataLoaders
- @eyurtsev

Twitter: [@UmerHAdil](https://twitter.com/@UmerHAdil) | Discord:
RicChilligerDude#7589

---------

Co-authored-by: UmerHA <40663591+UmerHA@users.noreply.github.com>

assert6 committed:
API update: Engines -> Models (#4915)

# API update: Engines -> Models

see: https://community.openai.com/t/api-update-engines-models/18597

Co-authored-by: assert <zhangchengming@kkguan.com>

eyurtsev committed:
feat #4479: TextLoader auto detect encoding and improved exceptions (#4927)

# TextLoader auto detect encoding and enhanced exception handling

- Add an option to enable encoding detection on `TextLoader`. 
- The detection is done using `chardet`
- The loading is done by trying all detected encodings by order of
confidence or raise an exception otherwise.

### New Dependencies:
- `chardet`

Fixes #4479 

## Before submitting

<!-- If you're adding a new integration, include an integration test and
an example notebook showing its use! -->

## Who can review?

Community members can review the PR once tests pass. Tag
maintainers/contributors who might be interested:

- @eyurtsev

---------

Co-authored-by: blob42 <spike@w530>

yuekaizhang committed:
Fix bilibili (#4860)

# Fix bilibili api import error

bilibili-api package is depracated and there is no sync module.

<!--
Thank you for contributing to LangChain! Your PR will appear in our next
release under the title you set. Please make sure it highlights your
valuable contribution.

Replace this with a description of the change, the issue it fixes (if
applicable), and relevant context. List any dependencies required for
this change.

After you're done, someone will review your PR. They may suggest
improvements. If no one reviews your PR within a few days, feel free to
@-mention the same people again, as notifications can get lost.
-->

<!-- Remove if not applicable -->

Fixes #2673 #2724 

## Before submitting

<!-- If you're adding a new integration, include an integration test and
an example notebook showing its use! -->

## Who can review?

Community members can review the PR once tests pass. Tag
maintainers/contributors who might be interested:
@vowelparrot  @liaokongVFX 

<!-- For a quicker response, figure out the right person to tag with @

        @hwchase17 - project lead

        Tracing / Callbacks
        - @agola11

        Async
        - @agola11

        DataLoaders
        - @eyurtsev

        Models
        - @hwchase17
        - @agola11

        Agents / Tools / Toolkits
        - @vowelparrot
        
        VectorStores / Retrievers / Memory
        - @dev2049
        
 -->

richardyc committed:
Add human message as input variable to chat agent prompt creation (#4542)

# Add human message as input variable to chat agent prompt creation

This PR adds human message and system message input to
`CHAT_ZERO_SHOT_REACT_DESCRIPTION` agent, similar to [conversational
chat
agent](https://github.com/hwchase17/langchain/blob/7bcf238a1acf40aef21a5a198cf0e62d76f93c15/langchain/agents/conversational_chat/base.py#L64-L71).

I met this issue trying to use `create_prompt` function when using the
[BabyAGI agent with tools
notebook](https://python.langchain.com/en/latest/use_cases/autonomous_agents/baby_agi_with_agent.html),
since BabyAGI uses “task” instead of “input” input variable. For normal
zero shot react agent this is fine because I can manually change the
suffix to “{input}/n/n{agent_scratchpad}” just like the notebook, but I
cannot do this with conversational chat agent, therefore blocking me to
use BabyAGI with chat zero shot agent.

I tested this in my own project
[Chrome-GPT](https://github.com/richardyc/Chrome-GPT) and this fix
worked.

## Request for review
Agents / Tools / Toolkits
- @vowelparrot

hwchase17 committed:
add alias for model (#4553)

Co-authored-by: Dev 2049 <dev.dev2049@gmail.com>

hwchase17 commented on langchain/agents/mrkl/output_parser.py:
```diff
@@ -22,6 +22,9 @@ def parse(self, text: str) -> Union[AgentAction, AgentFinish]:
             r"Action\s*\d*\s*:[\s]*(.*?)[\s]*Action\s*\d*\s*Input\s*\d*\s*:[\s]*(.*)"
         )
         match = re.search(regex, text, re.DOTALL)
+        if not match:  # try again with another format as LLM output is not stable
```
this seems a bit separate? could we handle/discuss in a separate pr?

gengliangwang replied:
@hwchase17 Sure. I just removed it.
cc @skcoirz Please create another PR for the parser changes and explain why we need it.

hwchase17 reviewed (commented):
one comment around some part looking a bit out of place, but otherwise looks good to me! if you dont get around to it, we can also remove the commented bit and merge in tonight @dev2049

hwchase17 committed:
dont error on sql import (#4647)

this makes it so we dont throw errors when importing langchain when
sqlalchemy==1.3.1

we dont really want to support 1.3.1 (seems like unneccessary maintance
cost) BUT we would like it to not terribly error should someone decide
to run on it

leo-gan committed:
docs: compound ecosystem and integrations (#4870)

# Docs: compound ecosystem and integrations

**Problem statement:** We have a big overlap between the
References/Integrations and Ecosystem/LongChain Ecosystem pages. It
confuses users. It creates a situation when new integration is added
only on one of these pages, which creates even more confusion.
- removed References/Integrations page (but move all its information
into the individual integration pages - in the next PR).
- renamed Ecosystem/LongChain Ecosystem into Integrations/Integrations.
I like the Ecosystem term. It is more generic and semantically richer
than the Integration term. But it mentally overloads users. The
`integration` term is more concrete.
UPDATE: after discussion, the Ecosystem is the term.
Ecosystem/Integrations is the page (in place of Ecosystem/LongChain
Ecosystem).

As a result, a user gets a single place to start with the individual
integration.

Chae4ek committed:
Update GPT4ALL integration (#4567)

# Update GPT4ALL integration

GPT4ALL have completely changed their bindings. They use a bit odd
implementation that doesn't fit well into base.py and it will probably
be changed again, so it's a temporary solution.

Fixes #3839, #4628

elBarkey committed:
FIX: GPTCache cache_obj creation loop (#4827)

_get_gptcache method keep creating new gptcache instance, here's the fix

# Fix GPTCache cache_obj creation loop

Fixes #4830 

Co-authored-by: Dev 2049 <dev.dev2049@gmail.com>

dev2049 committed:
Redis kwargs fix (#4936)

cc @tylerhutcherson

dev2049 committed:
Update redis integration tests (#4937)

leo-gan committed:
docs supabase update (#4935)

# docs: updated `Supabase` notebook

- the title of the notebook was inconsistent (included redundant
"Vectorstore"). Removed this "Vectorstore"
- added `Postgress` to the title. It is important. The `Postgres` name
is much more popular than `Supabase`.
- added description for the `Postrgress`
- added more info to the `Supabase` description

gengliangwang committed:
address comment

verygoodsoftwarenotvirus committed:
Correct typo in APIChain example notebook (Farenheit -> Fahrenheit) (#4938)

Correct typo in APIChain example notebook (Farenheit -> Fahrenheit)

so2liu committed:
fix: error in gptcache example nb (#4930)

vishwa-rn committed:
Update custom_multi_action_agent.ipynb (#4931)

Updated the docs from 
"An agent consists of three parts:" to 
"An agent consists of two parts:" since there are only two parts in the
documentation

leo-gan committed:
docs: added `ecosystem/dependents` page (#4941)

# docs: added `ecosystem/dependents` page

Added `ecosystem/dependents` page. Can we propose a better page name?

leo-gan committed:
docs: vectorstores, different updates and fixes (#4939)

# docs: vectorstores, different updates and fixes

Multiple updates:
- added/improved descriptions
- fixed header levels
- added headers
- fixed headers

gengliangwang committed:
fix import

gengliangwang committed:
Revert "add memory support to spark sql agent"

This reverts commit 1f3e54f9d42cd360ba8ada8f030cbfa527847dd5.

gengliangwang committed:
update notebook

gengliangwang committed:
fix lint

blob42 committed:
Chatconv agent: output parser exception (#4923)

the output parser form chat conversational agent now raises
`OutputParserException` like the rest.

The `raise OutputParserExeption(...) from e` form also carries through
the original error details on what went wrong.

I added the `ValueError` as a base class to `OutputParserException` to
avoid breaking code that was relying on `ValueError` as a way to catch
exceptions from the agent. So catching ValuError still works. Not sure
if this is a good idea though ?

gengliangwang commented:
I just did a final check before merging. 
There is a bug in the memory support. I reverted it to make this first version simple and robust. Discuss with @skcoirz offline and he will create another PR for general support for Agents. 
I also verified by rerunning the notebook. It works great.

danielchalef committed:
Zep Retriever - Vector Search Over Chat History (#4533)

# Zep Retriever - Vector Search Over Chat History with the Zep Long-term
Memory Service

More on Zep: https://github.com/getzep/zep

Note: This PR is related to and relies on
https://github.com/hwchase17/langchain/pull/4834. I did not want to
modify the `pyproject.toml` file to add the `zep-python` dependency a
second time.

Co-authored-by: Daniel Chalef <daniel.chalef@private.org>

jarib committed:
Fix get_num_tokens for Anthropic models (#4911)

The Anthropic classes used `BaseLanguageModel.get_num_tokens` because of
an issue with multiple inheritance. Fixed by moving the method from
`_AnthropicCommon` to both its subclasses.

This change will significantly speed up token counting for Anthropic
users.

scafati98 committed:
NIT: Instead of hardcoding k in each definition, define it as a param above. (#2675)

Co-authored-by: Dev 2049 <dev.dev2049@gmail.com>
Co-authored-by: Davis Chase <130488702+dev2049@users.noreply.github.com>

hwchase17 commented on langchain/spark_sql.py:
```diff
@@ -0,0 +1,174 @@
+from __future__ import annotations
```
nit: lets move this to utilities

dev2049 replied:
(as in the whole file should go to langchain/utilities/spark_sql, not that specific line)

hwchase17 replied:
yeah that whole file

gengliangwang replied:
Make sense. This is done.

skcoirz committed:
[nit] Simplify Spark Creation Validation Check A Little Bit (#4761)

- simplify the validation check a little bit.
- re-tested in jupyter notebook.

Reviewer: @hwchase17

gengliangwang committed:
refactor

gengliangwang committed:
Merge remote-tracking branch 'upstream/master' into spark_sql_no_subsclass

gengliangwang committed:
fix lint

gengliangwang commented:
@hwchase17 @dev2049 @skcoirz Thanks for reviewing this!
